### ISSUE TITLE
Loss curve cliffs on synthetic data Induction Heads task - is this expected?

### ISSUE THREAD

**Issue by hrbigelow** [state-spaces/mamba#61]
I just wanted to report a few loss curves on independent training runs.  I notice an odd pattern of an initial plateau followed by a rapid fall in the loss curve (don't tell Yudkowski!).  Am curious if others have seen this, and how to understand it?

These are using synthetic data as used in Section 4.1.2 Dataset.  I used a smaller `d_model=16` rather than the 64 used in the paper.

Specifically, I [generated data](https://github.com/hrbigelow/mamba-recall/blob/90e591383f01089a5752fe006c6220236302ac9c/data.py#L26) as:

```
#  T = randomly selected (but non-induction) token
#  S = special 'induction token'
#  M = particular 'memory token' randomly selected
1.  Generate synthetic data as TTTTTSMTTT...SM
2.  Train on the single cross entropy loss using the first n-1 tokens and predicting the very last 'M'
```

```
# Model Architecture
batch=8
n_layer=2
n_vocab=16
learn_rate=1e-3
d_model=16
n_epoch=40  # see note below
epoch_sz=1024

# synthetic data length
train_len=256,
```

For reproducibility, these are using this command:

`python3 train.py run7 delta.svlog --d_model=16 --n_epoch=40 --epoch_sz=1024 --report_every=500`

from [mamba-recall repo](https://github.com/hrbigelow/mamba-recall)

Here are 7 independent training runs showing the initial plateau and then a sudden drop in cross-entropy.  Actually, I realized I am NOT re-using the synthetic data for each `epoch` - so the notion of an epoch here is meaningless.  Each training batch is freshly synthetized.

![image](https://github.com/state-spaces/mamba/assets/5355286/86b47974-bd25-4146-9aaa-2479fd26c296)

Here is a zoom-in showing some blips.

![image](https://github.com/state-spaces/mamba/assets/5355286/9734451b-7e8a-4c83-b2e7-d533b792b2ba)

**Comment by albertfgu** [comment]
Seems related to the [grokking](https://arxiv.org/abs/2201.02177) effect which is still not that well understood. Pretty common for synthetic tasks. Seems more related to the data/task than to the model, have you tried other baselines?

**Comment by hrbigelow** [comment]
Ahh thanks, I didn't know grokking was more common in synthetic tasks, that's interesting.

I have not tried any other scenarios just yet except for the `d_model=64` case which also exhibits this.

I mainly wanted to confirm that this isn't a sign of something wrong in my setup.  Thanks!

**Comment by albertfgu** [comment]
Yep, it's expected. IIRC in my induction heads experiment there was a pretty stark phase change from stuck at baseline to perfectly solving the task.

**Comment by hrbigelow** [comment]
Great, thanks for sharing that!
